sdk/communication/AzureCommunicationChat/Source/ChatClient.swift

```diff
@@ -330,4 +338,109 @@ public class ChatClient {
+    /// - Parameters:
+    ///   - deviceToken: APNS push token.
+    ///   - completionHandler: Success indicates request to register for notifications has been received.
+    public func startPushNotifications(
+        deviceToken: String,
+        completionHandler: @escaping (Result<HTTPResponse?, AzureError>) -> Void
+    ) {
+        guard !pushNotificationsStarted else {
+            completionHandler(.failure(AzureError.client("Push notifications already started.")))
+            return
+        }
+
+        pushNotificationClient.registrationId = registrationId
+        pushNotificationClient.startPushNotifications(deviceRegistrationToken: deviceToken) { result in
+            switch result {
+            case let .success(response):
+                self.pushNotificationsStarted = true
+                completionHandler(.success(response))
+            case let .failure(error):
+                self.options.logger
+                    .error("Failed to start push notifications with error: \(error.localizedDescription)")
+                completionHandler(.failure(AzureError.client("Failed to start push notifications", error)))
+            }
+        }
+    }
+
+    /// Stop push notifications.
+    /// - Parameter completionHandler: Success indicates push notifications have been stopped.
+    public func stopPushNotifications(
+        completionHandler: @escaping (Result<HTTPResponse?, AzureError>) -> Void
+    ) {
+        guard pushNotificationsStarted else {
+            completionHandler(.failure(AzureError.client("Push notifications are not enabled.")))
+            return
+        }
+
+        // Report an error if pushNotificationClient doesn't exist
+        pushNotificationClient.stopPushNotifications { result in
+            switch result {
+            case let .success(response):
+                self.pushNotificationsStarted = false
+                let refreshedRegistrationId = UUID().uuidString
+                self.registrationId = refreshedRegistrationId
+                self.pushNotificationClient.registrationId = refreshedRegistrationId
+                completionHandler(.success(response))
+            case let .failure(error):
+                self.options.logger
+                    .error("Failed to stop push notifications with error: \(error.localizedDescription)")
+                completionHandler(.failure(AzureError.client("Failed to stop push notifications", error)))
+            }
+        }
+    }
+
+    /// Handle the data payload for an incoming push notification.
+    /// - Parameters:
+    ///   - notification: The APNS push notification payload ( including "aps" and "data" )
+    ///   - completionHandler:The event handler to handle the PushNotificationEvent
+    public func handlePush(
+        notification: [AnyHashable: Any],
+        completionHandler: PushNotificationEventHandler
+    ) {
```

Comment:
Again, `completionHandler` isn't really a completion handler and this method does not make sense and causes a lot of internal implementation to be exposed as public.

Why are you not following the pattern already used for realtime notifications where you register and unregister closures for specific events? Reviewing that pattern, it does not seems to unnecessarily expose any internal implementation.

Reply:
This has been discussed in our meeting.